### PR TITLE
Plural Data Sources Support Pagination

### LINKED ISSUE #168: Ensure pagination support works great
Need to chase down 20+ page length calls and make sure pagination works correctly

### PR DESCRIPTION
**Pull request by marinsalinas**
Closes #168

### CODE REVIEW

**Comment by JonKohler** [comment]
You'll need to rebase against develop when you get a chance. Should be a quick rebase

File: client/v3/v3_service.go [hunk 1]
```diff
@@ -824,3 +830,230 @@ func (op Operations) UpdateVolumeGroup(UUID string, body *VolumeGroupInput) (*Vo
...
 }
+
+const itemsPerPage = int64(100)
+
+func hasNext(ri *int64) bool {
+	*ri = *ri - itemsPerPage
+	return *ri >= (0 - itemsPerPage)
+}
+
+//ListAllVM ...
```
Comment (JonKohler) [inline client/v3/v3_service.go]: do // ListAllVMs (do a space after comment) for consistency

File: client/v3/v3_service.go [hunk 2]
```diff
@@ -824,3 +830,230 @@ func (op Operations) UpdateVolumeGroup(UUID string, body *VolumeGroupInput) (*Vo
...
+func (op Operations) ListAllVM() (*VMListIntentResponse, error) {
+	entities := make([]*VMIntentResource, 0)
+
+	//make first request
```
Comment (JonKohler) [inline client/v3/v3_service.go]: do space after comment for consistency per hashi

File: client/v3/v3_service.go [hunk 3]
```diff
@@ -824,3 +830,230 @@ func (op Operations) UpdateVolumeGroup(UUID string, body *VolumeGroupInput) (*Vo
...
+			offset += itemsPerPage
+			log.Printf("[Debug] total=%d, remaining=%d, offset=%d len(entities)=%d\n", totalEntities, remaining, offset, len(entities))
+		}
+		resp.Entities = entities
+	}
+
+	return resp, nil
+}
+
+//ListAllSubnet ...
```
Comment (JonKohler) [inline client/v3/v3_service.go]: do space after comment for consistency per hashi

File: client/v3/v3_service.go [hunk 4]
```diff
@@ -824,3 +830,230 @@ func (op Operations) UpdateVolumeGroup(UUID string, body *VolumeGroupInput) (*Vo
...
+func (op Operations) ListAllSubnet() (*SubnetListIntentResponse, error) {
+	entities := make([]*SubnetIntentResource, 0)
+
+	//make first request
```
Comment (JonKohler) [inline client/v3/v3_service.go]: do space after comment for consistency per hashi

File: client/v3/v3_service.go [hunk 5]
```diff
@@ -824,3 +830,230 @@ func (op Operations) UpdateVolumeGroup(UUID string, body *VolumeGroupInput) (*Vo
...
+				Kind:   utils.String("subnet"),
+				Length: utils.Int64(itemsPerPage),
+				Offset: utils.Int64(offset),
+			})
+
+			if err != nil {
+				return nil, err
+			}
+
+			//append to entities array
```
Comment (JonKohler) [inline client/v3/v3_service.go]: do space after comment for consistency per hashi

File: client/v3/v3_service.go [hunk 6]
```diff
@@ -824,3 +830,230 @@ func (op Operations) UpdateVolumeGroup(UUID string, body *VolumeGroupInput) (*Vo
...
+			offset += itemsPerPage
+			log.Printf("[Debug] total=%d, remaining=%d, offset=%d len(entities)=%d\n", totalEntities, remaining, offset, len(entities))
+		}
+		resp.Entities = entities
+	}
+
+	return resp, nil
+}
+
+//ListAllNetworkSecurityRule ...
```
Comment (JonKohler) [inline client/v3/v3_service.go]: do space after comment for consistency per hashi

File: client/v3/v3_service.go [hunk 7]
```diff
@@ -824,3 +830,230 @@ func (op Operations) UpdateVolumeGroup(UUID string, body *VolumeGroupInput) (*Vo
...
+func (op Operations) ListAllNetworkSecurityRule() (*NetworkSecurityRuleListIntentResponse, error) {
+	entities := make([]*NetworkSecurityRuleIntentResource, 0)
+
+	//make first request
```
Comment (JonKohler) [inline client/v3/v3_service.go]: do space after comment for consistency per hashi

**Comment by marinsalinas** [comment]
Request Changes Fixed!

**Comment by marinsalinas** [comment]
The test failures are fixed on delete-fix-#171 branch.

**Pull request merged**